src/shape/shape.js

```diff
@@ -113,44 +113,85 @@ extend(ChartInternal.prototype, {
 		};
 	},
 
-	getShapeOffset(typeFilter, indices, isSub) {
+	/**
+	 * @typedef {object} ShapeOffsetTarget
+	 * @property {string} id - target id
+	 * @property {object[]} rowValues - data point for each row (scaled as necessary)
+	 * @property {object<number, object>} rowValueMapByXValue - each x value is a key,
+	 *   mapped to the rowValue for that x value
+   * @property {number[]} values - value for each rowValue (normalized as necessary)
+	 */
+
+	/**
+	 * @param {function(Object): boolean} typeFilter
+	 * @return {{shapeOffsetTargets: ShapeOffsetTarget[], indexMapByTargetId: object}}
+	 */
+	getShapeOffsetData(typeFilter) {
 		const $$ = this;
 		const targets = $$.orderTargets($$.filterTargetsToShow($$.data.targets.filter(typeFilter, $$)));
-		const targetIds = targets.map(t => t.id);
+		const shapeOffsetTargets = targets.map(target => {
+			let rowValues = target.values;
+
+			if ($$.isStepType(target)) {
+				rowValues = $$.convertValuesToStep(rowValues);
+			}
+			const rowValueMapByXValue = rowValues.reduce((out, value) => {
+				out[Number(value.x)] = value;
+				return out;
+			}, {});
+
+			let values;
+
+			if ($$.isStackNormalized()) {
+				values = rowValues.map(v => $$.getRatio("index", v, true));
+			} else {
+				values = rowValues.map(({value}) => value);
```

Comment:
It doesn't necessarily to be called within iteration.
Declaring map function outside the loop, will improve performance little bit more.

```js
const getValues = $$.isStackNormalized() ?
	v => $$.getRatio("index", v, true) : ({value}) => value;

const shapeOffsetTargets = targets.map(target => {
	...
	const values = rowValues.map(getValues);
```

Reply:
I suppose this could provide a performance improvement, but I imagine it would be negligible. The `if/else` that your suggestion removes is only checked for each `target` - I'm not sure how many targets most people have, but in our case we have 10 on our busiest charts.

In my opinion, this is less readable, so I'd argue it's better to keep it as-is instead of sacrificing readability for 10 loops. If you feel otherwise, I can refactor.

Reply:
IMO the basic approach on every write is not doing unnecessary work if is possible.
Isn't a big deal, so feel free on this. Is okay as is.